### PR TITLE
win_iis_webapppool module, fixes and improvement

### LINKED ISSUE #2072: win_iis_webapppool failing with 'Specified cast is not valid'
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

win_iis_webapppool
##### ANSIBLE VERSION

```
ansible --version
ansible 2.0.0.2
  config file = 
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

No changes
##### OS / ENVIRONMENT

N/A
##### SUMMARY

The win_iis_webappool documentation states that you can separate attributes by a pipe and name values by colon like  "foo:1|bar:2"  However, doing this in the task results in a 'Specified cast is not valid error.'
##### STEPS TO REPRODUCE

Create a basic playbook with attributes that are separated by a pipe:

```
- name: Setting up IIS application pool 
    win_iis_webapppool: 
      name: PoolParty
      state: started 
      attributes: "managedRuntimeVersion:v4.0|managedPipelineMode:0|queueLength:30000"
```

##### EXPECTED RESULTS

App pool would get created successfully.
##### ACTUAL RESULTS

```
TASK [Setting up IIS application pool] *****************************************
fatal: [52.39.68.141]: FAILED! => {"changed": false, "failed": true, "msg": "Specified cast is not valid."}
```

### PR DESCRIPTION
**Pull request by jhawkesworth**
##### ISSUE TYPE

 - Bugfix Pull Request
 

##### COMPONENT NAME

win_iis_webapppool
##### ANSIBLE VERSION

```
ansible 2.3.0 (apppool-fixing a4c1ff3004) last updated 2017/01/26 08:23:24 (GMT +100)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```

##### SUMMARY

This change allows application pool properties which are based on controlled values in the iis configuration xml schema to be set and retrieved correctly, without turning off strict mode.

In order to retrieve the controlled values, the module now imports the
C:\Windows\system32\inetsrv\Microsoft.Web.Administration.dll
Further logic is required to retrieve the correct value for these controlled settings.
Also the enumeration names have to match the schema exactly, meaning the first letter of the property name has to be upper cased.

I also included the controlled (enumeration) values in the result object, allowing users to see which options are available.
The result object now returns the controlled name, rather than its array index position. For example instead of
"managedPipelineMode": 0,
the module now returns
"managedPipelineMode": "Classic",

Fixes https://github.com/ansible/ansible-modules-extras/issues/2072
and https://github.com/ansible/ansible/issues/20342

```
BEFORE:
ansible slweb -i DEPLOYTEST -l malpdwfftsla001 -m win_iis_webapppool -a "name='DefaultAppPool'"
Thursday 20 October 2016  16:11:18 -0400 (0:00:00.056)       0:00:00.056 ******
malpdwfftsla001 | SUCCESS => {
    "attributes": {},
    "changed": false,
    "info": {
        "attributes": {
            "CLRConfigFile": "",
            "applicationPoolSid": "S-1-5-82-3006700770-424185619-1745488364-794895919-4004696415",
            "autoStart": true,
            "enable32BitAppOnWin64": true,
            "enableConfigurationOverride": true,
            "managedPipelineMode": 0,
            "managedRuntimeLoader": "webengine4.dll",
            "managedRuntimeVersion": "v2.0",
            "name": "DefaultAppPool",
            "passAnonymousToken": true,
            "queueLength": 1000,
            "startMode": 0,
            "state": 1
        },
        "name": "DefaultAppPool",
        "state": "Started"
    }
}
AFTER:

ansible slweb -i DEPLOYTEST -l malpdwfftsla001 -m win_iis_webapppool -a "name='DefaultAppPool'" 

malpdwfftsla001 | SUCCESS => {
    "attributes": {},
    "changed": false,
    "info": {
        "attributes": {
            "CLRConfigFile": "",
            "applicationPoolSid": "S-1-5-82-3006700770-424185619-1745488364-794895919-4004696415",
            "autoStart": true,
            "enable32BitAppOnWin64": true,
            "enableConfigurationOverride": true,
            "managedPipelineMode": "Integrated",
            "managedRuntimeLoader": "webengine4.dll",
            "managedRuntimeVersion": "v2.0",
            "name": "DefaultAppPool",
            "passAnonymousToken": true,
            "queueLength": 1000,
            "startMode": "OnDemand",
            "state": "Started"
        },
        "enumerations": {
            "managedPipelineMode": [
                "Integrated",
                "Classic"
            ],
            "startMode": [
                "OnDemand",
                "AlwaysRunning"
            ],
            "state": [
                "Starting",
                "Started",
                "Stopping",
                "Stopped",
                "Unknown"
            ]
        },
        "name": "DefaultAppPool",
        "state": "Started"
    }
}
```

### CODE REVIEW

**Comment by jhawkesworth** [comment]
Replacement for PR I originally raised in extras https://github.com/ansible/ansible-modules-extras/pull/3257
which I believe fixes the problems with this module without resorting to turning off strict mode.

mentioning here
@michelleperz - who reported #2072
@deekayen - who opened #2578
@rriggs - who reports ansible/ansible#12494 (long since closed)

@nitzmahone - as requested on https://github.com/ansible/ansible/pull/20366 here's the recreated PR.

File: lib/ansible/modules/windows/win_iis_webapppool.ps1 [hunk 1]
```diff
@@ -49,6 +48,8 @@ If (Get-Member -InputObject $params -Name attributes) {
 # Ensure WebAdministration module is loaded
 if ((Get-Module "WebAdministration" -ErrorAction SilentlyContinue) -eq $NULL){
   Import-Module WebAdministration
+  Add-Type -Path C:\Windows\system32\inetsrv\Microsoft.Web.Administration.dll
```
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_iis_webapppool.ps1]: Minor nit: the path should probably be `Join-Path $env:SystemRoot system32\inetsrv\Microsoft.Web.Administration.dll` for those special snowflake systems with non C:\Windows system roots...

File: lib/ansible/modules/windows/win_iis_webapppool.ps1 [hunk 2]
```diff
@@ -90,18 +97,10 @@ try {
       Stop-WebAppPool -Name $name -ErrorAction Stop
       $result.changed = $TRUE
     }
-    if ((($state -eq 'started') -and ($pool.State -eq 'Stopped'))) {
+    if ((($state -eq 'started') -and ($pool.State -eq 'Stopped')) -or ($state -eq 'restarted')) {
```
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_iis_webapppool.ps1]: IIRC the separate `Restart-WebAppPool` call was required to get an actual restart- have you verified that's not the case? (docs are silent on the matter)

File: lib/ansible/modules/windows/win_iis_webapppool.ps1 [hunk 3]
```diff
@@ -115,9 +114,23 @@ if ($pool)
     name = $pool.Name
     state = $pool.State
     attributes =  New-Object psobject @{}
+    enumerations =  New-Object psobject @{}
```
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_iis_webapppool.ps1]: IIRC these values don't differ across hosts/OS/versions, right? Including the name translation in the output is great, but not sure of the utility of the `enumerations` retval over just snapshotting and including in docs...

File: lib/ansible/modules/windows/win_iis_webapppool.py [hunk 4]
```diff
@@ -58,40 +58,118 @@
 '''
 
 EXAMPLES = r'''
-- name: Return information about an existing application pool
-  win_iis_webapppool:
-    name: DefaultAppPool
+# This returns information about an existing application pool
```
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_iis_webapppool.py]: I think you and @Fale are fighting here- IIRC he just went and removed a bunch of the cmdline-style examples from modules in favor of play/task-style docs. I'd tend to agree with showing playbook examples of these things over direct cmdline samples- unless there's a caveat to k=v conversion of a particular module arg, it sounds like most folks prefer looking at the YAML style. That's the direction we've been going with a lot of stuff lately...

**Review by nitzmahone — DISMISSED** [review]
@jhawkesworth A couple of things to think about- I'll be out all next week, but if you guys want to duke it out here, IRC, whatever, feel free to merge or ping one of the other core folks to do it once everything is settled. Thanks!

File: lib/ansible/modules/windows/win_iis_webapppool.py (hunk 4 above)
Comment (Fale) [reply lib/ansible/modules/windows/win_iis_webapppool.py]: Yes I did, and that is actually what the [Documenti page](http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html) says ("Examples should be written in YAML format in plain text in an EXAMPLES string within the module")

**Comment by jhawkesworth** [comment]
looks like I have messed up my branch while attempting to fix.  Sigh.  Will probably try again with a third PR if I can't get my head round what I've messed up in git.

**Comment by jhawkesworth** [comment]
Ok, branch fixed.

@nitzmahone - all good points.  Fixed up the dll path and the key=value style examples.  I was attempting to get a bit of future proofing with 'enumerations' stuff, but having looked at what is in IIS 10, - https://www.iis.net/learn/get-started/whats-new-in-iis-10/new-features-introduced-in-iis-100 - not only is there a REST api to configure it, there are a bunch of much more targetted powershell Cmdlets, so it looks like there will be better options for future IIS versions anyway.  So it took out the enums from the module return anyway.

ready_for_review

**Comment by jhawkesworth** [comment]
bot_status

**Comment by mattclay** [comment]
Closing and re-opening to trigger CI.

File: lib/ansible/modules/windows/win_iis_webapppool.py [hunk 5]
```diff
@@ -47,51 +47,130 @@
...
     default: null
-    aliases: []
   attributes:
     description:
-      - Application Pool attributes from string where attributes are separated by a pipe and attribute name/values by colon Ex. "foo:1|bar:2"
+      - Application Pool attributes from string where attributes are separated by a pipe and attribute name/values by colon Ex. "foo:1|bar:2".
+      - The following attributes may only have the following names.
+      - managedPipelineMode may be either "Integrated" or  "Classic".
+      - startMode may be either "OnDemand" or  "AlwaysRunning".
+      - state may be one of "Starting", "Started", "Stopping", "Stopped", "Unknown".  Use the C(state) module parameter to modify, states shown are reflect the possible runtime values.
```
Comment (mattclay) [inline lib/ansible/modules/windows/win_iis_webapppool.py]: Ci failure due to PEP 8 issue:

```
2017-02-03 21:35:01 ERROR: PEP 8: lib/ansible/modules/windows/win_iis_webapppool.py:56:161: E501 line too long (184 > 160 characters) (current)
```

**Comment by jhawkesworth** [comment]
sorry, hadn't twigged about the line length check.  I'll fix asap.

**Comment by jhawkesworth** [comment]
closing and opening in the hope this triggers CI

**Comment by jhawkesworth** [comment]
ready_for_review

**Comment by jhawkesworth** [comment]
bot_status

**Comment by jhawkesworth** [comment]
Hello, sorry its taken so long to get this fix straight.
Would be great if anyone could check this over / try it out before ansible 2.3 feature freeze.
Mentioning a few people who might have an interest so hopefully you'll get a ping from github.

@michelleperz - who reported #2072
@deekayen - who opened #2578
@rriggs - who reports ansible/ansible#12494 (long since closed)
@jborean93 - don't know if you use IIS but if so this might be of interest to you too.

**Comment by jborean93** [comment]
Hmm I haven't had much experience with iis but I will see if I can find a way to test it before the deadline.

**Comment by jhawkesworth** [comment]
many thanks @jborean93 - there are some examples of installing IIS on http://docs.ansible.com/ansible/win_feature_module.html
Once you have it installed I think there's a default app pool created so you should be able to use the module against your (test) machine.

**Comment by jborean93** [comment]
@jhawkesworth just had a quick test with a 2016 and 2012 server and the output attributes for `state`, `startMode` and `managedPipelineMode` are coming out as integers instead of a human friendly name as shown by the documentation.

This is what I am getting;

```
TASK [debug] ***********************************************************************************************************
task path: /mnt/c/dev/test/role/tests_win_iis_webapppool.yml:7
ok: [server2012] => {
    "actual": {
        "attributes": {},
        "changed": false,
        "info": {
            "attributes": {
                "CLRConfigFile": "",
                "applicationPoolSid": "S-1-5-82-3006700770-424185619-1745488364-794895919-4004696415",
                "autoStart": true,
                "enable32BitAppOnWin64": false,
                "enableConfigurationOverride": true,
                "managedPipelineMode": 0,
                "managedRuntimeLoader": "webengine4.dll",
                "managedRuntimeVersion": "v4.0",
                "name": "DefaultAppPool",
                "passAnonymousToken": true,
                "queueLength": 1000,
                "startMode": 0,
                "state": 1
            },
            "name": "DefaultAppPool",
            "state": "Started"
        }
    },
    "changed": false
}
ok: [server2016] => {
    "actual": {
        "attributes": {},
        "changed": false,
        "info": {
            "attributes": {
                "CLRConfigFile": "",
                "applicationPoolSid": "S-1-5-82-3006700770-424185619-1745488364-794895919-4004696415",
                "autoStart": true,
                "enable32BitAppOnWin64": false,
                "enableConfigurationOverride": true,
                "managedPipelineMode": 0,
                "managedRuntimeLoader": "webengine4.dll",
                "managedRuntimeVersion": "",
                "name": "DefaultAppPool",
                "passAnonymousToken": true,
                "queueLength": 1000,
                "startMode": 0,
                "state": 1
            },
            "name": "DefaultAppPool",
            "state": "Started"
        }
    },
    "changed": false
}
```

**Comment by jborean93** [comment]
sorry to be the bearer of bad new @jhawkesworth but I am getting an error straight away :(

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: +                                                                   ~
fatal: [server2012]: FAILED! => {"changed": false, "failed": true, "msg": "At C:\\Users\\Administrator\\AppData\\Local\\Temp\\ansible-tmp-1486718237.31-167008817292242\\win_iis_webapppool.ps1:416 char:67\r\n+     if ((($state -eq 'started') -and ($pool.State -eq 'Stopped')) {\r\n+                                                                   ~\nUnexpected token '{' in expression or statement.\r\n\r\nAt C:\\Users\\Administrator\\AppData\\Local\\Temp\\ansible-tmp-1486718237.31-167008817292242\\win_iis_webapppool.ps1:416 char:67\r\n+     if ((($state -eq 'started') -and ($pool.State -eq 'Stopped')) {\r\n+                                                                   ~\nMissing closing ')' after expression in 'if' statement.\r\n\r\n"}
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: +                                                                   ~
fatal: [server2016]: FAILED! => {"changed": false, "failed": true, "msg": "At C:\\Users\\Administrator\\AppData\\Local\\Temp\\ansible-tmp-1486718237.31-209418506350668\\win_iis_webapppool.ps1:416 char:67\r\n+     if ((($state -eq 'started') -and ($pool.State -eq 'Stopped')) {\r\n+                                                                   ~\nUnexpected token '{' in expression or statement.\r\n\r\nAt C:\\Users\\Administrator\\AppData\\Local\\Temp\\ansible-tmp-1486718237.31-209418506350668\\win_iis_webapppool.ps1:416 char:67\r\n+     if ((($state -eq 'started') -and ($pool.State -eq 'Stopped')) {\r\n+                                                                   ~\nMissing closing ')' after expression in 'if' statement."}
```

**Comment by jhawkesworth** [comment]
thanks @jborean93 I'll dig in

**Comment by jhawkesworth** [comment]
fix pushed

ready_for_review

**Comment by jborean93** [comment]
Just one more question, there seems to be 2 attributes in the return value with the one not under `info` being empty, is this normal behaviour?

```
"actual": {
    "attributes": {}, # This is the one I am talking about
    "changed": false,
    "info": {
        "attributes": {
            "CLRConfigFile": "",
            "applicationPoolSid": "S-1-5-82-3006700770-424185619-1745488364-794895919-4004696415",
            "autoStart": true,
            "enable32BitAppOnWin64": false,
            "enableConfigurationOverride": true,
            "managedPipelineMode": "Integrated",
            "managedRuntimeLoader": "webengine4.dll",
            "managedRuntimeVersion": "v4.0",
            "name": "DefaultAppPool",
            "passAnonymousToken": true,
            "queueLength": 1000,
            "startMode": "OnDemand",
            "state": "Started"
        },
        "name": "DefaultAppPool",
        "state": "Started"
    }
},
"changed": false
````

**Comment by jhawkesworth** [comment]
Thanks for this. looks like I have more fixing to do, I will investigate
further.

**Comment by jborean93** [comment]
Looking at the module before your change it returns the same empty map so it doesn't look like something you have done. I'll do some more testing but it seems like the module should be cleaned up sometime in the future (for another time though :))
```
"actual": {
    "attributes": {},
    "changed": false,
    "info": {
        "attributes": {
            "CLRConfigFile": "",
            "applicationPoolSid": "S-1-5-82-3006700770-424185619-1745488364-794895919-4004696415",
            "autoStart": true,
            "enable32BitAppOnWin64": false,
            "enableConfigurationOverride": true,
            "managedPipelineMode": 0,
            "managedRuntimeLoader": "webengine4.dll",
            "managedRuntimeVersion": "v4.0",
            "name": "DefaultAppPool",
            "passAnonymousToken": true,
            "queueLength": 1000,
            "startMode": 0,
            "state": 1
        },
        "name": "DefaultAppPool",
        "state": "Started"
    }
}
```

**Comment by jborean93** [comment]
@jhawkesworth were you still looking into the attributes side for this module or are you ok with how it is?

**Comment by jhawkesworth** [comment]
@jborean93 I just pushed a change which I think *should* populate the top 'attributes'.
I also took the chance to tidy up, remove some ; and switch to Get-AnsibleParam and a hashtable for $results.
I can't test right now but if you have the chance to give it a go that would be very helpful.

**Comment by jhawkesworth** [comment]
ok pushed again, hopefully this time CI will accept the examples now I have removed the example output (which is probably a bit spurious now that we document RETURN values).

ready_for_review

**Comment by jborean93** [comment]
@jhawkesworth looks good to me, can confirm they are being populated. I think the module probably needs a tidy up in terms of formatting to match the other roles but we can leave that for another time.

shipit

**Comment by jhawkesworth** [comment]
bot_status

**Comment by jhawkesworth** [comment]
I need 1 more review here - can anyone help push this over the finish line and confirm it is working for them?

**Comment by deekayen** [comment]
Please merge it.

Here's my test playbook:

```
- hosts: all

  tasks:
    - name: "Install IIS."
      win_feature:
        name: "Web-Server"
        state: present
        restart: no
        include_sub_features: no
        include_management_tools: yes

    - name: "Create application pool with defaults."
      win_iis_webapppool:
        name: "GitHubPull20680_basic"
        state: started
      register: win_iis_webapppool_defaults

    - debug:
        var: win_iis_webapppool_defaults

    - name: "Create application pool with alternate settings."
      win_iis_webapppool:
        name: "GitHubPull20680_changes"
        state: started
        attributes: "autoStart:false|processModel.identityType:SpecificUser|processModel.userName:{{ user_iis_username }}|processModel.password:{{ user_iis_password }}|managedRuntimeVersion:v2.0|enable32BitAppOnWin64:true|managedPipelineMode:Classic"
      register: win_iis_webapppool_changes

    - debug:
        var: win_iis_webapppool_changes
```

The play output:

```
TASK [Create application pool with defaults.] **********************************
Monday 27 February 2017  17:13:56 -0500 (0:00:03.327)       0:00:14.491 ******* 
ok: [alpdpracticev01]

TASK [debug] *******************************************************************
Monday 27 February 2017  17:13:59 -0500 (0:00:03.016)       0:00:17.507 ******* 
ok: [alpdpracticev01] => {
    "win_iis_webapppool_defaults": {
        "attributes": {}, 
        "changed": false, 
        "info": {
            "attributes": {
                "CLRConfigFile": "", 
                "applicationPoolSid": "S-1-5-82-2845532531-1234513578-2234105921-2210852024-1136530100", 
                "autoStart": true, 
                "enable32BitAppOnWin64": false, 
                "enableConfigurationOverride": true, 
                "managedPipelineMode": "Integrated", 
                "managedRuntimeLoader": "webengine4.dll", 
                "managedRuntimeVersion": "v4.0", 
                "name": "GitHubPull20680_basic", 
                "passAnonymousToken": true, 
                "queueLength": 1000, 
                "startMode": "OnDemand", 
                "state": "Started"
            }, 
            "name": "GitHubPull20680_basic", 
            "state": "Started"
        }
    }
}

TASK [Create application pool with alternate settings.] ************************
Monday 27 February 2017  17:13:59 -0500 (0:00:00.089)       0:00:17.597 ******* 
changed: [alpdpracticev01]

TASK [debug] *******************************************************************
Monday 27 February 2017  17:14:07 -0500 (0:00:08.201)       0:00:25.798 ******* 
ok: [alpdpracticev01] => {
    "win_iis_webapppool_changes": {
        "attributes": {
            "autoStart": "false", 
            "enable32BitAppOnWin64": "true", 
            "managedPipelineMode": "Classic", 
            "managedRuntimeVersion": "v2.0", 
            "processModel.identityType": "SpecificUser", 
            "processModel.password": "hahathisisntreal", 
            "processModel.userName": "domain\\svc_account"
        }, 
        "changed": true, 
        "info": {
            "attributes": {
                "CLRConfigFile": "", 
                "applicationPoolSid": "S-1-5-82-2090956821-1234997103-1055353093-2979566528-3870201199", 
                "autoStart": false, 
                "enable32BitAppOnWin64": true, 
                "enableConfigurationOverride": true, 
                "managedPipelineMode": "Classic", 
                "managedRuntimeLoader": "webengine4.dll", 
                "managedRuntimeVersion": "v2.0", 
                "name": "GitHubPull20680_changes", 
                "passAnonymousToken": true, 
                "queueLength": 1000, 
                "startMode": "OnDemand", 
                "state": "Started"
            }, 
            "name": "GitHubPull20680_changes", 
            "state": "Started"
        }
    }
}
```

The result was as I expected:

screen_shot_2017-02-27_at_5_15_34_pm

**Comment by jhawkesworth** [comment]
Thanks for testing @deekayen 

Looks like it has been tested by two people now, so I will ask about getting this merged.

**Comment by virtuald** [comment]
I have the latest ansible installed (from git branch stable-2.3), and this fix isn't working for me, as I can't set arbitrary parameters. I tried the play that @deekayen did, and it doesn't seem to be setting the parameters correctly. I'm on Windows Server 2012. I'd love to debug if someone could give me some pointers.

**Comment by jhawkesworth** [comment]
@virtuald - probably best to open a new issue as closed Pull Requests tend not to get monitored.
Please ensure you have run 'source hacking/env-setup' and make sure you aren't picking up any old versions of win_iis_webapppool.ps1 (run a find to see if you have multiple ones).  If that doesn't fix it, please open a new issue with a minimal reproducer playbook included.  many thanks, Jon

**Comment by virtuald** [comment]
Bug filed in #22805

**Pull request merged**